File: zenscript-code-model/src/main/java/raylras/zen/model/type/MapEntryType.java
package raylras.zen.model.type;

import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;
import raylras.zen.model.symbol.SymbolProvider;

import java.util.List;

public record MapEntryType(Type keyType, Type valueType) implements Type, SymbolProvider {

    @Override
    public String getTypeName() {
        return "Map.Entry<" + keyType.getTypeName() + "," + valueType.getTypeName() + ">";
    }

    @Override
    public String getSimpleTypeName() {
        return "Map.Entry<" + keyType.getSimpleTypeName() + "," + valueType.getSimpleTypeName() + ">";
    }

    @Override
    public List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .variable("key", keyType, Symbol.Modifier.IMPLICIT_VAL)
                .variable("value", valueType, Symbol.Modifier.IMPLICIT_VAL)
                .build();
    }

    @Override
    public String toString() {
        return getTypeName();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/MapEntryType.kt

package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.SymbolProvider
import raylras.zen.model.symbol.symbolSequence

data class MapEntryType(val keyType: Type, val valueType: Type) : Type, SymbolProvider {
    override val typeName by lazy { "Map.Entry<${keyType.typeName},${valueType.typeName}>" }

    override val simpleTypeName by lazy { "Map.Entry<${keyType.simpleTypeName},${valueType.simpleTypeName}>" }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        return symbolSequence {
            variable("key", keyType)
            variable("value", valueType)
        }
    }

    override fun toString() = typeName
}
